TestingLowerBounds/ForMathlib/LeftRightDeriv.lean

```diff
@@ -60,6 +60,65 @@ lemma leftDeriv_eq_rightDeriv (f : ℝ → ℝ) :
   ext x
   simp [leftDeriv_eq_rightDeriv_apply]
 
+@[simp]
+lemma rightDeriv_const (c : ℝ) : rightDeriv (fun _ ↦ c) = 0 := by
+  ext x
+  rw [rightDeriv_def, Pi.zero_apply]
+  exact derivWithin_const x _ c (uniqueDiffWithinAt_Ioi x)
+
+@[simp]
+lemma leftDeriv_const (c : ℝ) : leftDeriv (fun _ ↦ c) = 0 := by
+  simp_rw [leftDeriv_eq_rightDeriv, rightDeriv_const, Pi.zero_apply, neg_zero]
+  rfl
+
+lemma rightDeriv_linear (a : ℝ) : rightDeriv (fun x ↦ a * x) = fun _ ↦ a := by
+  ext x
+  rw [rightDeriv_def, derivWithin_const_mul (uniqueDiffWithinAt_Ioi x)]
+  swap; · exact differentiableWithinAt_id
+  change a * derivWithin id (Ioi x) x = a
+  rw [derivWithin_id _ _ (uniqueDiffWithinAt_Ioi x), mul_one]
+
+lemma leftDeriv_linear (a : ℝ) : leftDeriv (fun x ↦ a * x) = fun _ ↦ a := by
+  simp_rw [leftDeriv_eq_rightDeriv, ← neg_mul_comm, rightDeriv_linear]
+  ext; simp
+
+lemma rightDeriv_add {f g : ℝ → ℝ} (hf : ∀ x, DifferentiableWithinAt ℝ f (Ioi x) x)
+    (hg : ∀ x, DifferentiableWithinAt ℝ g (Ioi x) x) :
+    rightDeriv (f + g) = fun x ↦ rightDeriv f x + rightDeriv g x := by
```

Comment:
You should have a second version of this, in the form
`rightDeriv (fun x ↦ f x + g x) = fun x ↦ rightDeriv f x + rightDeriv g x`.
Then you wont need the `change` lines in all lemmas below.

Reply:
Also you should write those lemmas for one particular `x` and replace `hf` by `(hf : DifferentiableWithinAt ℝ f (Ioi x) x)`. That way if the differentiability hypothesis is true only on a subset, you can still use the lemma for that subset (that issue came up during the derivAtTop refactor).